### ISSUE TITLE
Feature Request: Pause/Resume timers

### ISSUE THREAD

**Issue by Floyz** [babybuddy/babybuddy#190]
Hello :)
Sometimes i'd like to pause/resume a timer because baby is doing a quick wake up for quick feed or change and i'd like to pause/resume the sleep track (without stop/create another one).
Maybe it's already available but i haven't seen the option.
Regards !

**Comment by cdubz** [comment]
Yes, I agree this would be useful at times. The core models using timers do record duration separately from start and end time so in theory this is possible without core model data structure changes but it would still take a considerable amount of refactoring to support pause/resume behavior. This isn't something I would prioritize right now but I'd be open to a PR.

**Comment by Munksgaard** [comment]
:+1: for this feature request, would be very helpful.

**Comment by lordratner** [comment]
Agreed, this would be very useful for accurately tracking feedings

**Comment by cdubz** [comment]
I am still open to this. An approach would need to start with some research around a proper data model. The current timer model has these relevant fields: https://github.com/babybuddy/babybuddy/blob/f5e9d4dacde102950aa6611eab4dca2dd7d0ddf2/core/models.py#L543-L551

When the timer is stopped and saved, the `duration` is calculated: https://github.com/babybuddy/babybuddy/blob/f5e9d4dacde102950aa6611eab4dca2dd7d0ddf2/core/models.py#L609-L616

So I _think_ a rudimentary way to approach this would be to add two new fields:

1. A `pause` field with the same signature as `end`.
2. A `paused_duration` field with the same signature as `duration` (except perhaps defaulting to zero instead of null).

And do a bit of refactoring:

- Add a `Timer.pause` method to set the current datetime to the `pause` value.
- Add a `Timer.resume` method to add the difference of the current datetime minus `pause` to `paused_duration` and clear the value of `pause`.
- Refactor `Timer.save` to calculate `duration` as the difference between `start` and `end` _minus_ `paused_duration`.
- Surely some other refactoring to account for a new state somewhere along the line.

I think that model would support keeping the original `start` datetime and allowing for any number of pause and resume operations without much friction.

Does this miss anything? Would it still capture the most relevant data? Or does it loss anything the current model gets us?

Is there already a more well known/tested model for this somewhere we can draw from?

**Comment by lordratner** [comment]
I really like how ParentLove does it a lot. Here's the breastfeeding timer:

![Screenshot_20230228-235249](https://user-images.githubusercontent.com/49489398/222056712-cafc8ec3-4c1f-40e2-a3ee-2e144f1fdaa9.png)

Each side can be run independently, or concurrently, and paused as many times as needed. It keeps track of the number of pauses and the duration of pauses, but they don't show up on the timer page, only the history page. You end up with the cumulative time for each breast at the end of the session, you hit save, and it's committed to the history:

![Screenshot_20230228-235926](https://user-images.githubusercontent.com/49489398/222057679-c1195366-d8c7-478a-9a36-68db8afb3b8e.png)

You can see the history shows the start time, end time, and feeding time per breast. It also shows the number of pauses, and if you open up the entry it shows how long the pauses were:

![Screenshot_20230228-235845](https://user-images.githubusercontent.com/49489398/222057896-a1384f9a-250d-4695-a7a3-7a232a69a6bd.png)

Within an individual history entry, you can edit the start time, the end time, the date, and the cumulative time for each breast. You can also select which breast was last used. Here's what it looks like changing the cumulative times:

![Screenshot_20230301-000253](https://user-images.githubusercontent.com/49489398/222058141-27ba2340-3917-4f5d-8adc-b444da2fbc8b.png)

**Comment by lordratner** [comment]
I forgot to upload these

Here's the timer running:
![Screenshot_20230301-000710](https://user-images.githubusercontent.com/49489398/222059026-c5e900fc-704c-4f99-9238-6fcf51c32afb.png)


And here's the timer paused:
![Screenshot_20230301-000720](https://user-images.githubusercontent.com/49489398/222059080-e09541c8-7744-4da0-bd2f-48eb33e7d9ef.png)

**Comment by cdubz** [comment]
@lordratner thanks for sharing.

That is a good deal more complex than the basic pause/resume behavior requested here. I think it may make more more sense as a separate feature request (unless there isn't much real value in the simpler approach).

One follow up question re: the ParentLove timer -- how do you use the data? It has been a while since we had the need for this app and back then out understanding of the importance of this was mostly just around total timing as a proxy for "amount" of feeding for a child. I'm specifically interested in how you use and act on the per-breast timing level of detail.

**Comment by lordratner** [comment]
> @lordratner thanks for sharing.
> 
> That is a good deal more complex than the basic pause/resume behavior requested here. I think it may make more more sense as a separate feature request (unless there isn't much real value in the simpler approach).
> 
> One follow up question re: the ParentLove timer -- how do you use the data? It has been a while since we had the need for this app and back then out understanding of the importance of this was mostly just around total timing as a proxy for "amount" of feeding for a child. I'm specifically interested in how you use and act on the per-breast timing level of detail.

For the per-breast timing, my wife tries to keep the feeding balanced. So for a 30 minute feed she'll do 15 on each side. Sometimes that's 15left 15 right, other times it's 10 left, 15 right, 5 left. It just helps her make sure she is working both sides equally to prevent soreness and encourage production. I think. It's also helpful to her to know which breast was last used, so she can start with the other. Of course we are only on day 3, so who knows how this will change?

I was thinking of a way to do this, and I think I have something that would allow for initially keeping the right/left timers combined, then easily adding the feature to have two breast timers in one overall timer. However it would probably require some changes to how the timer type is determined. Right now you start the timer, and when you select the type it ends the timer. I'm sure there's a way to keep that dynamic, but it might be simpler to just pick the timer type right when you start it, that way the timer can have the additional functionality (which breast, pause unpause) from the start. I envision when you press the "+" in the upper right to start the timer, a dropdown with the timer types appears, then when you pick the type it goes to the appropriate timer page.

As for the "how" lets just start with a single timer for both breasts, as currently implemented.

You'd need these variables:

start  -- same as current

end -- same as current

is_paused -- true/false, changed by pressing the pause button

pauses = [] This is an array of tuples. each tuple is two parts and appended to the array when the pause button is pushed. The first is a timestamp and the second is either 1 for paused or -1 for unpaused. Might look like this after starting, then a few pauses and unpauses: 

```
[
(1677686206, 1),
(1677686216,-1),
(1677686293,1),
(1677686380,1-)
]
```
The 1 or -1 makes it very simple to add or subtract the time without any additional coding. 


The duration can now be calculated by iterating through the array 

```
duration = now() #or end if the timer is already complete
for i in pauses:
    duration = duration + (pauses[i][0] * pauses [i][1])
duration = duration - start
```

Duration can be calculated at anytime while the timer is running, and when the timer is stopped for good duration can be calculated and saved, then the array is dumped. Or you can keep the array if you want to play with something later. 

This could pretty easily be used to track both breasts in a later revision. You'd still have a start and end variable, but you'd have is_paused_left and is_paused_right

Then you'd have pauses_left and pauses_right arrays as described above, each button would append to the applicable array. 

This is obviously just the math of making it work. I'm terrible with methods (and python in general) so you'd have to fit it into that construct, but this shows how duration could be calculated and tracked in a very lightweight manner. You only need to track the timestamps when the pause or unpause occurred, rather than each duration. 

Does that make sense?

**Comment by cdubz** [comment]
Yes, this makes sense. Although it also makes me feel like we would just a separate model specifically for a breast feeding timer. I can't think of any enhancements we'd bring to the other timers (sleep, tummy time) with a predefined timer-type. And the breast feeding timer could also be reusable as a pumping timer (which doesn't have a quick timer button yet anyway).

So I think the pausing functionality I still like my approach in https://github.com/babybuddy/babybuddy/issues/190#issuecomment-1449341929 and could see the rest of this issue broken out a separate feature request.

**Comment by kisamoto** [comment]
Another possibility could be to aggregate/group durations to a timer. This would be a bit of a refactor with the `start`, `stop` being pulled out into a `Duration` model with a foreign key back to a `Timer`.

A start event would create a new timer and a new duration. No pauses = 1 Duration row. For each pause a Duration has the `stop` time added (optional pre-calculate the duration). When resumed, a new row with the current `start` time is added and repeat.

The `Timer` can have a total running duration calculated as the latest `stop` minus the earliest `start` time and an active duration as a sum of all related `Duration` objects. 

Not sure that makes complete sense so here's a simplified example:

A timer table:

id|name|
--|----|
1|`foo bar`

A duration table with multiple durations representing pause & resumes (could optionally add a precomputed `duration` column as exists today):

id|timer_id|start|stop|duration
--|--|--|--| --
1 | 1| 2024-09-11T15:07:07.315Z | 2024-09-11T15:15:07.315Z |
2 | 1| 2024-09-11T15:30:07.315Z | 2024-09-11T16:00:07.315Z | 

then pseudocode for calculating durations would be:

1. Total duration for timer `1` = `(latest stop for duration - earliest start for duration) where timer id = 1`
2. Total active duration = `sum duration where timer id = 1`
